src/tpm.rs

```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
+    let (ret_out, _) = run("tpm2_nvlist".to_string(), None)?;
+    let ret_out_value: Value = serde_yaml::from_str(&ret_out)?;
+    if ret_out_value["0x1c00002"]["size"].is_null() {
+        return Err(KeylimeTpmError::new_tpm_rust_error(
+            "TPM nvlist failed.",
+        ));
+    }
+
+    let ekcert_size = ret_out_value["0x1c00002"]["size"].clone();
+    let (_, f_out) = run(
+        format!("tpm2_nvread -x 0x1c00002 -s {} -f {}", ekcert_size, nvpath),
+        Some(vec![nvpath]),
+    )?;
+    let ekcert = f_out.get(nvpath).ok_or_else(|| {
+        KeylimeTpmError::new_tpm_rust_error(
+            "failed to read nvram and retrieve ekcert.",
+        )
+    })?;
+    Ok(base64::encode(ekcert))
+}
+
+/*
+ *  Return:
+ *     value from nvram
+ *     Error if the process failed
+ *
+ *  Read the data stored in a Non-Volatile (NV)s index.
+ */
+fn tpm_read_key_nvram() -> Result<String, KeylimeTpmError> {
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let (ret_out, _) = run(
+        format!(
+            "tpm2_nvread -x 0x1500018 -a 0x40000001 -s {} -P {}",
+            common::BOOTSTRAP_KEY_SIZE,
+            owner_pw
+        ),
+        None,
+    )?;
+
+    if ret_out.len() as i32 == common::BOOTSTRAP_KEY_SIZE {
+        return Ok(ret_out.into());
+    }
+    Ok(String::new())
+}
+
+/*
+ *  Input:
+ *      key to write to nvram
+ *  Return:
+ *      Error if process failed
+ *
+ *  Write data to a Non-Volatile (NV) index.
+ */
+fn tpm_write_key_nvram(key: String) -> Result<(), KeylimeTpmError> {
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let mut tf = NamedTempFile::new()?;
+    tf.write_all(key.as_bytes())?;
+    let attrs = "ownerread|policywrite|ownerwrite";
+    let tf_path = temp_file_get_path(&tf)?;
+
+    run(format!("tpm2_nvdefine -x 0x1500018 -a 0x40000001 -s {} -t \"{}\" -p {} -P {}", common::BOOTSTRAP_KEY_SIZE, attrs, owner_pw, owner_pw), None)?;
+    run(
+        format!(
+            "tpm2_nvwrite -x 0x1500018 -a 0x40000001 -P {} {}",
+            owner_pw, tf_path
+        ),
+        None,
+    )?;
+    Ok(())
+}
+
+/*
+ *  Input:
+ *      activation flag
+ *      asymmetric algorithm
+ *      hash algorithm
+ *      sign algorithm
+ *  Return:
+ *      Error if the process failed.
+ *
+ *  Create a new aik key from TPM. Store the aik_name, aik_password,
+ *  aik_public_key and aik_handle to tpmdata.json file.
+ */
+fn tpm_create_aik(
+    activate: bool,
+    asym_alg: String,
+    hash_alg: String,
+    sign_alg: String,
+) -> Result<(), KeylimeTpmError> {
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let aik = get_tpm_metadata_content("aik")?;
+    let aik_name = get_tpm_metadata_content("aik_name")?;
+
+    if !aik.is_empty() && !aik_name.is_empty() {
+        let aik_handle = get_tpm_metadata_content("aik_handle")?;
+
+        let (ret_out, _) =
+            run("tpm2_getcap -c handles-persisitent".to_string(), None)?;
```

Comment:
This is incorrectly spelt: `handles-persisitent` will fail, should be:

`run("tpm2_getcap -c handles-persistent".to_string(), None)?;`